apps/wallet/src/app/core/services/ledger.service.ts

```
@@ -0,0 +1,286 @@
+      transport: this.transport,
+    };
+  }
+
+  public async connect(): Promise<void> {
+    this.ledger = new WavesLedger(this.ledgerOptions);
+
+    // wait until device is connected
+    const ledgerData = await this.ledger.getUserDataById(this.userId, false);
+    this.userData = {
+      ...ledgerData,
+      name: `Ledger Wallet ${this.userId}`,
+    };
+
+    this.connected$.next(true);
+  }
+
+  public async disconnect(): Promise<void> {
+    if (!this.ledger) throw new Error('Ledger not connected');
+
+    await this.ledger.disconnect();
+    this.connected$.next(false);
+  }
+
+  public async signAndBroadcast(data: IUnsignedTransaction): Promise<void> {
+    if (!this.ledger) throw new Error('Ledger not connected');
+    if (!this.userData) throw new Error('Error with Ledger address data');
+
+    let schema, version, rawTransaction, prefixBytes;
+    let unsignedTransaction: ITransaction & WithId;
+    const senderPublicKey = this.userData.publicKey;
+
+    switch (data.type) {
+      case TransactionTypes.TRANSFER:
+        if (!data.recipient) throw new Error('Property "recipient" is undefined');
+        if (!data.amount) throw new Error('Property "amount" is undefined');
+        if (!data.attachment) data.attachment = '';
+
+        // @todo: for now, ledger signing only works with transfer v1, need to add more recent versions to it (v2, v3)
+        // https://github.com/iicc1/ledger-app-lto/issues/3
+        version = 1;
+
+        rawTransaction = {
+          version,
+          fee: data.fee,
+          senderPublicKey,
+          amount: data.amount,
+          recipient: data.recipient,
+          attachment: data.attachment,
+          type: data.type as unknown as TRANSACTION_TYPE.TRANSFER,
+        };
+
+        unsignedTransaction = transfer(rawTransaction);
+
+        prefixBytes = new Uint8Array();
+        schema = parseSerialize.schemas.transferSchemaV1;
+        break;
+      case TransactionTypes.MASS_TRANSFER:
```

If those won't be supported, I suggest comment them out (security concerns)?

---

What are the security concerns?

---

Can this method be accessed from an API call? If so when someone makes tx with such type, we should output the default case: `throw new Error('Unknown transaction type')` (or even better - 'Not supported transaction type') and not put it down the line and mess up things eventually?